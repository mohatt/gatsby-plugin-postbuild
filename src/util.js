import path from 'path'
import crypto from 'crypto'
import _ from 'lodash'
import Debug from 'debug'

/**
 * Global options object
 *
 * @type {Object}
 */
export let options = {
  _plugin: 'gatsby-plugin-postbuild'
}

/**
 * Initializes plguin options and other utilities
 *
 * @param {Object} $0
 * @param   {Object} $0.defaultOptions - Options defaults
 * @param   {Object} $0.pluginOptions - Plugin options
 * @param   {Object} $0.gatsby - Gatsby's onPreBootstrap event object
 */
export function bootstrap ({ defaultOptions, pluginOptions, gatsby }) {
  // Merge user-defined options with defaults
  options = _.mergeWith(
    options,
    defaultOptions,
    pluginOptions,
    // Ensure arrays aren't merged by index
    (to, from) => _.isArray(from) ? from : undefined
  )
  // set private options
  options._root = gatsby.store.getState().program.directory
  options._public = path.join(options._root, 'public')
  options._pathPrefix = gatsby.pathPrefix

  // Initialize reporter
  const reporter = gatsby.reporter
  const errorMap = {
    10000: {
      text: context => `${options._plugin} threw an error while running:\n ${context.message}`,
      level: 'ERROR',
      type: 'PLUGIN'
    }
  }
  if (reporter.setErrorMap) {
    reporter.setErrorMap(errorMap)
  }
  debug('Plugin initialized', options)
}

/**
 * Creates a new structured error to pass to gatsby's reporter
 *
 * @constructor
 * @param {string|Object} message
 * @param {Error|null} error
 * @param {number} code
 * @return {{ id: string, context: {Object}, error: error}}
 */
export function ReporterError (message, error = null, code = 10000) {
  return {
    id: `${options._plugin}_${code}`,
    context: typeof message === 'object'
      ? message
      : { message },
    error: error && typeof error === 'object'
      ? error
      : undefined
  }
}

/**
 * Prints a debug message
 *
 * @param {string} message
 * @param {any} data
 */
export function debug (message, data = null) {
  createDebug()(...arguments)
}

/**
 * Creates a debug instance for the given namespace
 *
 * @param {string} namespace
 * @return {Function}
 */
export function createDebug (namespace = '') {
  let ns = options._plugin
  if (namespace) {
    ns += ':' + namespace
  }

  createDebug.cache ??= new Map()
  if (createDebug.cache.has(ns)) {
    return createDebug.cache.get(ns)
  }

  const realDebug = Debug(ns)
  const debug = function (message) {
    arguments[0] = arguments.length > 1
      ? `${message} %O \n\n`
      : `${message} \n\n`
    realDebug(...arguments)
  }

  createDebug.cache.set(ns, debug)
  return debug
}

/**
 * Creates a sha1 hash from a string.
 *
 * @param {string} data
 * @param {boolean} base64
 */
export function sha1 (data, base64 = false) {
  return crypto.createHash('sha1')
    .update(data, 'utf8')
    .digest(base64 ? 'base64' : 'hex')
}
