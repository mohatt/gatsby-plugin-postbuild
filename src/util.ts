import path from 'path'
import crypto from 'crypto'
import { mergeWith, isArray } from 'lodash'
import Debug from 'debug'
import { defaults, IPluginOptions } from './options'
import {
  GatsbyNodeHelpers,
  GatsbyPluginOptions,
  GatsbyReporterErrorMap,
  GatsbyReporterErrorMeta
} from './gatsby'

/**
 * Options interface
 */
export interface IOptions extends IPluginOptions {
  _plugin: string
  _root: string
  _public: string
  _pathPrefix: string
}

/**
 * Global options object
 */
export let options: IOptions = {
  _plugin: 'gatsby-plugin-postbuild',
  _root: '',
  _public: '',
  _pathPrefix: '',
  ...defaults
}

/**
 * Initializes plguin options and other utilities
 */
export function bootstrap ({ pluginOptions, gatsby }: {
  pluginOptions: GatsbyPluginOptions
  gatsby: GatsbyNodeHelpers
}): void {
  // Merge user-defined options with defaults
  options = mergeWith(options, pluginOptions,
    // Ensure arrays aren't merged by index
    (to, from) => isArray(from) ? from : undefined
  )

  // set private options
  options._root = gatsby.store.getState().program.directory
  options._public = path.join(options._root, 'public')
  options._pathPrefix = gatsby.pathPrefix

  // Initialize reporter
  const reporter = gatsby.reporter
  const errorMap: GatsbyReporterErrorMap = {
    10000: {
      text: (context: { message: string }) => {
        return `${options._plugin} threw an error while running:\n ${context.message}`
      },
      level: 'ERROR',
      type: 'PLUGIN'
    }
  }
  if (typeof reporter.setErrorMap === 'function') {
    reporter.setErrorMap(errorMap)
  }

  // Debug final options
  debug('Plugin initialized', options)
}

/**
 * Creates a structured error object to pass to gatsby's reporter
 */
export function createGatsbyError (message: string|Object, error?: Error, code = 10000): GatsbyReporterErrorMeta {
  return {
    id: `${options._plugin}_${code}`,
    context: typeof message === 'object'
      ? message
      : { message },
    error: error instanceof Error
      ? error
      : undefined
  }
}

/**
 * Prints a debug message
 */
export function debug (message: string, ...params: any[]): void {
  createDebug()(message, ...params)
}

/**
 * Creates a debug instance for the given namespace
 *
 * @param {string} namespace
 * @return {Function}
 */
export function createDebug (namespace: string = ''): Function {
  let ns = options._plugin
  if (namespace !== '') {
    ns += ':' + namespace
  }

  const debug = Debug(ns)
  return function (message: string, ...params: any[]) {
    message = params.length > 0
      ? `${message} %O \n\n`
      : `${message} \n\n`
    debug(message, ...params)
  }
}

/**
 * Creates a sha1 hash from a string.
 */
export function sha1 (data: string, base64 = false): string {
  return crypto.createHash('sha1')
    .update(data, 'utf8')
    .digest(base64 ? 'base64' : 'hex')
}

/**
 * Returns the file extension or checks if
 * a file has a given extension
 */
export function extName (file: string, checkExt?: string): string|boolean {
  const ext = path.extname(file).replace('.', '').toLowerCase()
  if (checkExt === undefined) return ext
  return ext === checkExt
}
