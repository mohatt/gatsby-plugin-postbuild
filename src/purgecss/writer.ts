import { promises as fs } from 'fs'
import path from 'path'
import filesize from 'filesize'
import { options, createDebug } from '../util'
const debug = createDebug('purgecss/writer')

/**
 * Represents a file written by the plugin
 * along with some stats
 */
export interface IWriteReport {
  file: string
  sizes: string[]
  rejected?: number
}

/**
 * Handles writing optimized files
 */
export class FileWriter {
  /**
   * Formats a given file size
   */
  formatSize: ((bytes: number) => string)

  /**
   * File write reports
   */
  reports: IWriteReport[] = []

  /**
   * Total bytes of data purged
   */
  saving: number = 0

  /**
   * Creates the filesize format function
   */
  constructor () {
    this.formatSize = filesize.partial({ spacer: '' })
  }

  /**
   * Writes the purged file to disk
   *
   * @param file - File absolute path
   * @param data - File data
   * @param purged - Purged selectors as array
   */
  async write (file: string, data: string, purged: string[]): Promise<void> {
    const sizes = []
    try {
      // ensure we're not creating any new files
      await fs.access(file)
      // Caclulate data lengths
      sizes.push((await fs.stat(file)).size)
      sizes.push(Buffer.byteLength(data))
      sizes.push(sizes[1] - sizes[0])
      this.saving += sizes[2]
      // Write the actual optimized file
      debug('Writing purged file', file)
      await fs.writeFile(file, data)
      // Write purged log if enabled
      if (options.reportRejected) {
        await fs.writeFile(file + '.rejected.log', purged.join(' '))
      }
    } catch (e) {
      throw new Error(`Unable to write purged file "${file}": ${String(e.message)}`)
    }

    const r = {
      file: path.relative(options._public, file),
      sizes: sizes.map(this.formatSize),
      rejected: purged.length
    }
    this.reports.push(r)
    // Ignore reporting to console if disabled
    if (!options.reportConsole) {
      return
    }

    // Write report to console
    console.log(
      (this.reports.length === 1 ? '\n' : '') +
      `\x1b[33m ${r.file}\x1b[0m\x1b[2m` +
      `\n  |- Size: ${r.sizes[0]} -> ${r.sizes[1]} (${r.sizes[2]})` +
      `\n  |- Rejected: ${r.rejected} selector${r.rejected !== 1 ? 's' : ''}` +
      '\x1b[0m\n'
    )
  }

  /**
   * Returns a list of all write reports
   */
  getReports (): IWriteReport[] {
    return this.reports
  }

  /**
   * Returns the total bytes saved and its formatted form
   */
  getTotalSaving (): [number, string] {
    const saving = this.saving * -1
    return [
      saving,
      this.formatSize(saving)
    ]
  }
}
