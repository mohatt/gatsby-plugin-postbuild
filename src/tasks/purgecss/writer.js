import { promises as fs } from 'fs'
import path from 'path'
import filesize from 'filesize'
import { options } from '../../util'

export class FileWriter {
  formatSize
  reports = []
  saving = 0

  constructor () {
    this.formatSize = filesize.partial({ spacer: '' })
  }

  /**
   * Writes the purged asset file on disk
   *
   * @param {string} file - File absolute path
   * @param {string} data - File data
   * @param {([string]|null)} purged - Purged selectors as array to write a rejected
   *  log file next to the main asset file, otherwise nothing will be written
   * @return {Object} - A summary report to be printed
   */
  async write (file, data, purged) {
    const sizes = []
    const rejected = Array.isArray(purged)
    try {
      await fs.access(file)
      // Caclulate data lengths
      sizes.push((await fs.stat(file)).size)
      sizes.push(Buffer.byteLength(data))
      sizes.push(sizes[1] - sizes[0])
      this.saving += sizes[2]
      // Write the actual file
      await fs.writeFile(file, data)
      // Write purged log if enabled
      if (rejected) {
        await fs.writeFile(file + '.rejected.log', purged.join(' '))
      }
    } catch (e) {
      throw new Error(`Unable to write purged file "${file}": ${e.message}`)
    }

    const r = {
      file: path.relative(options._public, file),
      sizes: sizes.map(this.formatSize),
      rejected: rejected ? purged.length : undefined
    }
    this.reports.push(r)
    // Ignore reporting to console if specified
    if (!options.reportConsole) {
      return
    }

    // Write report to console
    console.log(`\x1b[33m
 ${r.file}\x1b[0m\x1b[2m
  |- Size: ${r.sizes[0]} -> ${r.sizes[1]} (${r.sizes[2]})` +
      (rejected
        ? `
  |- Rejected: ${r.rejected} selector${r.rejected !== 1 ? 's' : ''}`
        : '') +
      '\x1b[0m')
  }

  getReports () {
    return this.reports
  }

  getTotalSaving () {
    const saving = this.saving * -1
    return [
      saving,
      this.formatSize(saving)
    ]
  }
}
