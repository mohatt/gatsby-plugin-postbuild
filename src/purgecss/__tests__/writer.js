import fs from 'fs'
import { options } from '../../util'
import { mountFile, mountOptions } from '../../../test/utils'
import { FileWriter } from '../writer'

jest.mock('fs')
mountOptions()

describe('FileWriter', () => {
  const filename = `${options._public}/page/index.html`
  const data = '//ok'
  console.log = jest.fn()

  beforeEach(() => {
    console.log.mockReset()
    fs.reset()
  })

  it('correctly writes files', async () => {
    mountFile(filename, '//noop')
    mountOptions({ reportConsole: true })
    let writer
    expect(() => { writer = new FileWriter() }).not.toThrow()
    await expect(writer.write(filename, data)).resolves.not.toThrow()
    expect(fs.readFileSync(filename, 'utf-8')).toBe(data)
    expect(console.log).toHaveBeenCalledTimes(1)
    expect(writer.getReports()).toMatchSnapshot()
    expect(writer.getTotalSaving()).toMatchSnapshot()
  })

  it('correctly ignores reporting to console', async () => {
    mountOptions({ reportConsole: false })
    mountFile(filename, '//noop')
    let writer
    expect(() => { writer = new FileWriter() }).not.toThrow()
    await expect(writer.write(filename, data)).resolves.not.toThrow()
    expect(console.log).not.toHaveBeenCalled()
  })

  it('correctly ignores non-existing files', async () => {
    let writer
    expect(() => { writer = new FileWriter() }).not.toThrow()
    await expect(writer.write(filename, data)).rejects.toThrow()
  })

  it('correctly writes rejected log file', async () => {
    mountFile(filename, '//noop')
    let writer
    expect(() => { writer = new FileWriter() }).not.toThrow()
    await expect(writer.write(filename, data, ['a', 'b'])).resolves.not.toThrow()
    expect(fs.readFileSync(filename + '.rejected.log', 'utf-8')).toBe('a b')
  })
})
