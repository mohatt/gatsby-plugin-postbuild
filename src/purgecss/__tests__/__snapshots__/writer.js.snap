// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileWriter correctly writes files 1`] = `
Array [
  Object {
    "file": "page/index.html",
    "rejected": 0,
    "sizes": Array [
      "6B",
      "4B",
      "-2B",
    ],
  },
]
`;

exports[`FileWriter correctly writes files 2`] = `
Array [
  2,
  "2B",
]
`;
